win32/src/threadpool_win32.c

```diff
@@ -99,14 +100,38 @@ static void internal_close(THREADPOOL_HANDLE threadpool)
     /* Codes_SRS_THREADPOOL_WIN32_01_033: [ threadpool_close shall destroy the thread pool environment created in threadpool_open_async. ]*/
     DestroyThreadpoolEnvironment(&threadpool->tp_environment);
 
-    (void)InterlockedExchange(&threadpool->state, (LONG)THREADPOOL_WIN32_STATE_CLOSED);
+    (void)InterlockedExchange(&threadpool->state, THREADPOOL_WIN32_STATE_CLOSED);
     WakeByAddressSingle((PVOID)&threadpool->state);
 }
 
-THREADPOOL_HANDLE threadpool_create(EXECUTION_ENGINE_HANDLE execution_engine)
+static void threadpool_dispose(THREADPOOL* threadpool)
 {
-    THREADPOOL_HANDLE result;
+    /* Codes_SRS_THREADPOOL_WIN32_01_005: [ threadpool_dispose shall free all resources associated with threadpool. ] */
+    /* Codes_SRS_THREADPOOL_WIN32_01_006: [ While threadpool is OPENING or CLOSING, threadpool_dispose shall wait for the open to complete either successfully or with error. ]*/
+    do
+    {
+        int32_t current_state = InterlockedCompareExchange(&threadpool->state, THREADPOOL_WIN32_STATE_CLOSING, THREADPOOL_WIN32_STATE_OPEN);
+        if (current_state == THREADPOOL_WIN32_STATE_OPEN)
+        {
+            /* Codes_SRS_THREADPOOL_WIN32_01_007: [ threadpool_destroy shall perform an implicit close if threadpool is OPEN. ]*/
+            internal_close(threadpool);
+            break;
+        }
+        else if (current_state == THREADPOOL_WIN32_STATE_CLOSED)
+        {
+            break;
+        }
+
+        (void)WaitOnAddress(&threadpool->state, &current_state, sizeof(current_state), INFINITE);
+    } while (1);
 
+    /* Codes_SRS_THREADPOOL_WIN32_42_028: [ threadpool_dispose shall decrement the reference count on the execution_engine. ]*/
+    execution_engine_dec_ref(threadpool->execution_engine);
+}
+
+THANDLE(THREADPOOL) threadpool_create(EXECUTION_ENGINE_HANDLE execution_engine)
+{
+    THREADPOOL* result = NULL;

```

Comment:
>THREADPOOL [](http://example.com/codeflow?start=4&length=10)

THANDLE(THREADPOOL)

Reply:
THANDLE(Threadpool) will not build because it make this const and you have to initialize it here.  This needs to be THREADPOOL*

Reply:
but we are returning a THANDLE(THREADPOOL), so result should technically be that type. Typically we do this by assigning to NULL at declaration and using the THANDLE_GET_T macro as needed.

I mean, it will compile as-is, just looks a bit wrong to me because `THANDLE(THREADPOOL)` is actually a different type than `THREADPOOL*`